cantools/database/utils.py

```diff
@@ -66,49 +67,43 @@ def start_bit(data: Union["Data", "Signal"]) -> int:
         return data.start
 
 
-def _encode_fields(fields: Sequence[Union["Signal", "Data"]],
-                   data: SignalMappingType,
-                   scaling: bool,
-                   ) -> Dict[str, Union[int, float]]:
+def _encode_signals(signals: Sequence[Union["Signal", "Data"]],
+                    data: SignalMappingType,
+                    scaling: bool,
+                    ) -> Dict[str, Union[int, float]]:
     unpacked = {}
-    for field in fields:
-        value = data[field.name]
-
-        if isinstance(value, (float, int)):
-            _transform = float if field.is_float else round
-            if scaling:
-                if field.offset == 0 and field.scale == 1:
-                    # treat special case to avoid introduction of unnecessary rounding error
-                    unpacked[field.name] = _transform(value)  # type: ignore[operator]
-                    continue
-
-                unpacked[field.name] = _transform((value - field.offset) / field.scale)  # type: ignore[operator]
-                continue
-
-            unpacked[field.name] = _transform(value)  # type: ignore[operator]
-            continue
+    for signal in signals:
+        value = data[signal.name]
 
-        if isinstance(value, (str, NamedSignalValue)):
-            unpacked[field.name] = field.choice_string_to_number(str(value))
-            continue
+        raw_value: Union[int, float]
+        if scaling:
+            raw_value = signal.scaled_to_raw(value)
+        else:
+            if isinstance(value, (str, NamedSignalValue)):
```

Comment:
I think there is something wrong here. If `scaling` is False, then `value` must be the raw value. The `choice_to_number()` was already performed in `scaled_to_raw()`

Reply:
this is pretty twisted:

> If scaling is False, then value must be the raw value.

it can also be a choice (i.e., 'str' or 'NamedSignalValue'). it can be argued that considering choices should only be done if `scaling` is `True`, thouhgh. (the existing code does it unconditionally, so I thought I keep it this way...)